File: scripts/JavaScript/sudoku.js
let board = [
    [5, 0, 0, 0, 7, 0, 0, 1, 0],
    [0, 0, 4, 0, 0, 6, 0, 0, 9],
    [7, 0, 0, 0, 0, 0, 0, 6, 0],
    [6, 0, 0, 9, 5, 0, 0, 0, 0],
    [0, 5, 2, 6, 4, 8, 0, 0, 0],
    [8, 0, 0, 3, 0, 0, 0, 0, 0],
    [2, 0, 5, 8, 0, 0, 0, 4, 0],
    [1, 0, 8, 0, 0, 0, 0, 0, 7],
    [0, 9, 0, 0, 0, 0, 2, 5, 0]
]
function get_valids(i,j){
    var valids = [];
    var add = true;
    var ib = Math.floor(i/3)*3;
    var jb = Math.floor(j/3)*3;
    for (let num = 1;num<10;num++){
        add = true;
        for (let x = 0;x<9;x++){
            if (board[x][j] == num || board[i][x] == num){
                add = false;
                break;
            }
        }
        for (let g = ib;g<ib+3;g++){
            for (let h = jb;h<jb+3;h++){
                if (board[g][h] == num){
                    add = false;
                    break;
                }
            }
        }
        if (add == true){
            valids.push(num);
        }
    }
    return valids;
    
};
function solve_singels() {
    var prev_sudoku = board.slice()
    while (true){
        for (let i = 0; i <9; i++) {
            for (let j = 0; j <9; j++) {
                var valids = get_valids(i,j);
                if (valids.length == 1){
                    board[i][j] = valids[0];
                }
            }
        }
        if (board.every((val,index) => val === prev_sudoku[index])){
            return board;
        }else{
            prev_sudoku = board.slice();
        }
    }
};
function vec_od_nic(val) {
    return val!=0;
}
function solve(){
    if (board.every(vec_od_nic)){
        console.log("Solved");
        console.log(board);
        return
    }
    for(let i = 0;i<=9;i++){
        for(let j = 0;j<=9;j++){
            if (board[i][j] == 0){
                var valids = get_valids(i,j);
                if (valids.length != 0){
                    for(var valid of valids){
                        board[i][j] = valid;
                        solve();
                        board[i][j] = 0;
                    }
                }
                return;
            }
        }
    }
}
solve_singels(board);
solve();

File: scripts/python/Sudoku/sudoku.py
#!/usr/bin/python3
import numpy as np
import math
import time as t
import cv2
from video import get_sudoku
import os
digit_pos = {"1":(190,1755),"2":(380,1755),"3":(550,1755),"4":(750,1755),"5":(890,1755),"6":(190,1930),"7":(380,1930),"8":(550,1930),"9":(750,1930)}
def get_image(ime):
    command = f"adb exec-out screencap -p > {ime}.jpg"
    os.system(command)
    im = cv2.imread(f"{ime}.jpg")
    im = im[400:1480][0:1080]
    # cv2.imshow("title",im)
    cv2.imwrite(f"{ime}.jpg",im)
ime = "sudoTest1"
#   get_image(ime)
img = cv2.imread(f"{ime}.jpg")
sudoku = get_sudoku(img)
# print(sudoku)
def get_valids(sudoku, i, j):
    valids = []
    add = True
    ib = math.floor(i/3)*3
    jb = math.floor(j/3)*3
    for num in range(1,10):
        add = True
        for x in range(0,9):
            if sudoku[x][j] == num or sudoku[i][x] == num:
                add = False
                break
        for g in range(ib,ib+3):
            for h in range(jb,jb+3):
                if sudoku[g][h] == num:
                    add = False
                    break
        if add == True:
            valids.append(num)
    return valids
def solve_singels(sudoku):
    prev_sudoku = np.copy(sudoku)
    while True:
        for i in range(0, 9):
            for j in range(0, 9):
                if sudoku[i][j] == 0:
                    valids = get_valids(sudoku, i, j)
                    if len(valids) == 1:
                        sudoku[i][j] = valids[0]
        if np.allclose(prev_sudoku,sudoku):
            return sudoku
        else:
            prev_sudoku = np.copy(sudoku)
def solve():
    if np.all(sudoku != 0) == True:
        print("solved")
        to_app2(sudoku)
        return sudoku
    for i in range(0, 9):
        for j in range(0, 9):
            if sudoku[i][j] == 0:
                valids = get_valids(sudoku, i, j)
                if len(valids) != 0:
                    for valid in valids:
                        sudoku[i][j] = valid
                        del valid
                        solve()
                        sudoku[i][j] = 0
                return
def to_app(digit,pos1,pos2):
    command = "adb shell input tap "+str(digit_pos[digit][0])+" "+ str(digit_pos[digit][1])
    os.system(command)
    command2 = "adb shell input tap "+str(pos1)+" "+ str(pos2)
    os.system(command2)
def to_app2(sudoku):
    print(sudoku)
    for i in range(0,9):
        for j in range(0,9):
            # print(str(int(sudoku[i][j])),str(j*100+85),str(i*100+450))
            to_app(str(int(sudoku[i][j])),str(j*100+85),str(i*100+450))
            t.sleep(0.1)
solve()